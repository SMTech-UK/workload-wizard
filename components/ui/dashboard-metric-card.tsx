import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-500">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-muted-foreground mt-1">{subtitle}</div>}
    </div>
  );
} 