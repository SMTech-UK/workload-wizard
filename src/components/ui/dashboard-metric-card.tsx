import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white dark:bg-zinc-900 shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-900 dark:text-white">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-gray-600 dark:text-gray-300 mt-1">{subtitle}</div>}
    </div>
  );
} 