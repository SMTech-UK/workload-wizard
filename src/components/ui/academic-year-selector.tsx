"use client"

import React from "react";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
import { Badge } from "@/components/ui/badge";
import { useAcademicYear } from "@/hooks/useAcademicYear";
import { Calendar } from "lucide-react";

interface AcademicYearSelectorProps {
  className?: string;
  showActiveBadge?: boolean;
  onAcademicYearChange?: (academicYearId: string) => void;
}

export function AcademicYearSelector({ 
  className, 
  showActiveBadge = true,
  onAcademicYearChange 
}: AcademicYearSelectorProps) {
  const { 
    academicYears, 
    currentAcademicYear, 
    currentAcademicYearId, 
    changeAcademicYear 
  } = useAcademicYear();

  const handleAcademicYearChange = async (academicYearId: string) => {
    try {
      await changeAcademicYear(academicYearId);
      onAcademicYearChange?.(academicYearId);
    } catch (error) {
      console.error('Failed to change academic year:', error);
    }
  };

  // Function to format academic year name (e.g., "2025/26" -> "AY-25/26")
  const formatAcademicYearName = (yearName: string) => {
    // Extract the last 2 digits from each year part
    const parts = yearName.split('/');
    if (parts.length === 2) {
      const startYear = parts[0].slice(-2);
      const endYear = parts[1].slice(-2);
      return `AY-${startYear}/${endYear}`;
    }
    // Fallback to original name if format is unexpected
    return yearName;
  };

  if (!academicYears || academicYears.length === 0) {
    return (
      <div className={`flex items-center gap-2 text-muted-foreground ${className}`}>
        <Calendar className="h-4 w-4" />
        <span className="text-sm">No academic years configured</span>
      </div>
    );
  }

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      <Select
        value={currentAcademicYearId || ""}
        onValueChange={handleAcademicYearChange}
      >
        <SelectTrigger className="w-auto min-w-[120px] flex items-center gap-2">
          <Calendar className="h-4 w-4 text-muted-foreground" />
          <span className="font-medium">
            {currentAcademicYear ? formatAcademicYearName(currentAcademicYear.name) : "Select AY"}
          </span>
        </SelectTrigger>
        <SelectContent>
          {academicYears.map((year: any) => (
            <SelectItem key={year._id} value={year._id}>
              <div className="flex items-center gap-2">
                <Calendar className="h-4 w-4 text-muted-foreground" />
                <span>{formatAcademicYearName(year.name)}</span>
                {year.isActive && showActiveBadge && (
                  <Badge variant="default" className="text-xs">Active</Badge>
                )}
                {year.isStaging && (
                  <Badge variant="secondary" className="text-xs">Staging</Badge>
                )}
              </div>
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
    </div>
  );
} 