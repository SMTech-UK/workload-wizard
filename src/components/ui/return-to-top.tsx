"use client";

import { Button } from "@/components/ui/button";
import { ChevronUp, Home } from "lucide-react";
import { useState, useEffect } from "react";
import Link from "next/link";

export function ReturnToTop() {
  const [isVisible, setIsVisible] = useState(true); // Temporarily always visible for testing

  useEffect(() => {
    const toggleVisibility = () => {
      console.log("Scroll position:", window.pageYOffset);
      if (window.pageYOffset > 300) {
        console.log("Setting visible to true");
        setIsVisible(true);
      } else {
        console.log("Setting visible to false");
        setIsVisible(false);
      }
    };

    // Call once on mount to check initial state
    toggleVisibility();
    
    window.addEventListener("scroll", toggleVisibility);
    return () => window.removeEventListener("scroll", toggleVisibility);
  }, []);

  const scrollToTop = () => {
    console.log("Scrolling to top");
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };

  console.log("ReturnToTop component render, isVisible:", isVisible);

  return (
    <>
      {/* Return to Home button - top left */}
      <Link href="/">
        <Button
          size="sm"
          variant="outline"
          className="fixed top-4 left-4 z-50 shadow-lg hover:shadow-xl transition-all duration-200"
          aria-label="Return to home"
        >
          <Home className="h-4 w-4 mr-2" />
          Home
        </Button>
      </Link>

      {/* Return to Top button - bottom left */}
      <Button
        onClick={scrollToTop}
        size="sm"
        className="fixed bottom-4 left-4 z-50 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 bg-blue-500 text-white"
        aria-label="Return to top"
      >
        <ChevronUp className="h-4 w-4" />
      </Button>
    </>
  );
} 