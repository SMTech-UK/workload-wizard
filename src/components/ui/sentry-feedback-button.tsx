"use client";

import { useEffect, useState } from "react";
import * as Sentry from "@sentry/nextjs";
import { Button } from "@/components/ui/button";
import { Bug, MessageCircle } from "lucide-react";
import { useModalAwareButton } from "@/hooks/useModalAwareButton";
import { createPortal } from "react-dom";

export function SentryFeedbackButton() {
  const [feedback, setFeedback] = useState<any>(null);
  const [widget, setWidget] = useState<any>(null);
  const [mounted, setMounted] = useState(false);
  const buttonRef = useModalAwareButton();

  useEffect(() => {
    setMounted(true);
    // Get the feedback instance on the client side
    setFeedback(Sentry.getFeedback());
  }, []);

  const handleFeedbackClick = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    
    if (!feedback) return;

    if (widget) {
      // If widget is open, close it
      widget.removeFromDom();
      setWidget(null);
    } else {
      // Create and show the form immediately
      feedback.createForm().then((newWidget: any) => {
        newWidget.appendToDom();
        newWidget.open();
        setWidget(newWidget);
      }).catch((error: any) => {
        console.error('Failed to create Sentry feedback form:', error);
      });
    }
  };

  if (!feedback || !mounted) return null;

  const button = (
    <Button
      ref={buttonRef}
      variant="destructive"
      size="lg"
      onClick={handleFeedbackClick}
      onMouseDown={(e) => {
        e.preventDefault();
        e.stopPropagation();
      }}
      className="fixed bottom-4 right-4 z-[99999] shadow-lg"
      style={{
        zIndex: 99999,
        pointerEvents: 'auto',
      }}
      aria-label="Report a bug or give feedback"
    >
      <Bug className="h-4 w-4 mr-2" />
      Report an Issue
    </Button>
  );

  // Render the button in a portal to ensure it's outside modal context
  return createPortal(button, document.body);
} 