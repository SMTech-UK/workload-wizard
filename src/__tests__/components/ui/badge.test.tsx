import { render, screen } from '@testing-library/react';
import { Badge } from '@/components/ui/badge';

describe('Badge Component - Comprehensive UI Testing', () => {
  describe('Basic Rendering and Functionality', () => {
    it('should render badge with text content', () => {
      render(<Badge>Test Badge</Badge>);

      expect(screen.getByText('Test Badge')).toBeInTheDocument();
      expect(screen.getByText('Test Badge')).toHaveClass('inline-flex', 'items-center', 'rounded-md', 'border');
    });

    it('should render empty badge', () => {
      const { container } = render(<Badge />);

      const badge = container.querySelector('.inline-flex');
      expect(badge).toBeInTheDocument();
      expect(badge).toHaveClass('inline-flex', 'items-center', 'rounded-md', 'border');
    });

    it('should render badge with numbers', () => {
      render(<Badge>42</Badge>);

      expect(screen.getByText('42')).toBeInTheDocument();
    });

    it('should render badge with special characters', () => {
      render(<Badge>New!</Badge>);

      expect(screen.getByText('New!')).toBeInTheDocument();
    });
  });

  describe('Accessibility Testing', () => {
    it('should have proper semantic structure', () => {
      render(<Badge>Accessible Badge</Badge>);

      const badge = screen.getByText('Accessible Badge');
      expect(badge.tagName).toBe('DIV');
      expect(badge).toHaveClass('inline-flex', 'items-center', 'rounded-md', 'border');
    });

    it('should support focus states', () => {
      render(<Badge tabIndex={0}>Focusable Badge</Badge>);

      const badge = screen.getByText('Focusable Badge');
      expect(badge).toHaveAttribute('tabIndex', '0');
      expect(badge).toHaveClass('focus:outline-none', 'focus:ring-2', 'focus:ring-ring', 'focus:ring-offset-2');
    });

    it('should handle ARIA attributes', () => {
      render(<Badge aria-label="Status badge">Active</Badge>);

      const badge = screen.getByLabelText('Status badge');
      expect(badge).toBeInTheDocument();
      expect(badge).toHaveTextContent('Active');
    });
  });

  describe('Variant Testing', () => {
    it('should render default variant correctly', () => {
      render(<Badge variant="default">Default Badge</Badge>);

      const badge = screen.getByText('Default Badge');
      expect(badge).toHaveClass('border-transparent', 'bg-primary', 'text-primary-foreground', 'shadow');
      expect(badge).toHaveClass('hover:bg-primary/80');
    });

    it('should render secondary variant correctly', () => {
      render(<Badge variant="secondary">Secondary Badge</Badge>);

      const badge = screen.getByText('Secondary Badge');
      expect(badge).toHaveClass('border-transparent', 'bg-secondary', 'text-secondary-foreground');
      expect(badge).toHaveClass('hover:bg-secondary/80');
    });

    it('should render destructive variant correctly', () => {
      render(<Badge variant="destructive">Destructive Badge</Badge>);

      const badge = screen.getByText('Destructive Badge');
      expect(badge).toHaveClass('border-transparent', 'bg-destructive', 'text-destructive-foreground', 'shadow');
      expect(badge).toHaveClass('hover:bg-destructive/80');
    });

    it('should render outline variant correctly', () => {
      render(<Badge variant="outline">Outline Badge</Badge>);

      const badge = screen.getByText('Outline Badge');
      expect(badge).toHaveClass('text-foreground');
    });

    it('should use default variant when no variant is specified', () => {
      render(<Badge>No Variant Badge</Badge>);

      const badge = screen.getByText('No Variant Badge');
      expect(badge).toHaveClass('border-transparent', 'bg-primary', 'text-primary-foreground', 'shadow');
    });
  });

  describe('Responsive Behavior', () => {
    it('should maintain proper sizing on different screen sizes', () => {
      // Mock mobile screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 375,
      });

      render(<Badge>Mobile Badge</Badge>);

      const badge = screen.getByText('Mobile Badge');
      expect(badge).toHaveClass('px-2.5', 'py-0.5', 'text-xs');
    });

    it('should maintain proper sizing on tablet screens', () => {
      // Mock tablet screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 768,
      });

      render(<Badge>Tablet Badge</Badge>);

      const badge = screen.getByText('Tablet Badge');
      expect(badge).toHaveClass('px-2.5', 'py-0.5', 'text-xs');
    });

    it('should maintain proper sizing on desktop screens', () => {
      // Mock desktop screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1024,
      });

      render(<Badge>Desktop Badge</Badge>);

      const badge = screen.getByText('Desktop Badge');
      expect(badge).toHaveClass('px-2.5', 'py-0.5', 'text-xs');
    });
  });

  describe('Content Structure', () => {
    it('should handle long content gracefully', () => {
      const longContent = 'A'.repeat(100);

      render(<Badge>{longContent}</Badge>);

      const badge = screen.getByText(longContent);
      expect(badge).toBeInTheDocument();
      expect(badge).toHaveClass('inline-flex', 'items-center');
    });

    it('should handle special characters in content', () => {
      const specialContent = 'Badge with special chars: <>&"\'<>';

      render(<Badge>{specialContent}</Badge>);

      expect(screen.getByText(specialContent)).toBeInTheDocument();
    });

    it('should handle HTML content safely', () => {
      const htmlContent = '<strong>Bold</strong> and <em>italic</em>';

      render(<Badge>{htmlContent}</Badge>);

      expect(screen.getByText(htmlContent)).toBeInTheDocument();
    });

    it('should handle mixed content types', () => {
      render(<Badge>Text with 123 numbers and !@# symbols</Badge>);

      expect(screen.getByText('Text with 123 numbers and !@# symbols')).toBeInTheDocument();
    });
  });

  describe('Edge Cases and Error Handling', () => {
    it('should handle very long content', () => {
      const veryLongContent = 'A'.repeat(500);

      render(<Badge>{veryLongContent}</Badge>);

      const badge = screen.getByText(veryLongContent);
      expect(badge).toBeInTheDocument();
      expect(badge).toHaveClass('inline-flex', 'items-center');
    });

    it('should handle empty strings', () => {
      const { container } = render(<Badge></Badge>);

      const badge = container.querySelector('.inline-flex');
      expect(badge).toBeInTheDocument();
      expect(badge).toHaveClass('inline-flex', 'items-center', 'rounded-md', 'border');
    });

    it('should handle multiple badges', () => {
      render(
        <div>
          <Badge>First Badge</Badge>
          <Badge>Second Badge</Badge>
          <Badge>Third Badge</Badge>
        </div>
      );

      expect(screen.getByText('First Badge')).toBeInTheDocument();
      expect(screen.getByText('Second Badge')).toBeInTheDocument();
      expect(screen.getByText('Third Badge')).toBeInTheDocument();
    });

    it('should handle badges with custom className', () => {
      render(<Badge className="custom-class">Custom Badge</Badge>);

      const badge = screen.getByText('Custom Badge');
      expect(badge).toHaveClass('custom-class');
      expect(badge).toHaveClass('inline-flex', 'items-center', 'rounded-md', 'border');
    });
  });

  describe('Snapshot Testing for UI Consistency', () => {
    it('should match snapshot for default badge', () => {
      const { container } = render(<Badge>Default Badge</Badge>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for secondary badge', () => {
      const { container } = render(<Badge variant="secondary">Secondary Badge</Badge>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for destructive badge', () => {
      const { container } = render(<Badge variant="destructive">Destructive Badge</Badge>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for outline badge', () => {
      const { container } = render(<Badge variant="outline">Outline Badge</Badge>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for empty badge', () => {
      const { container } = render(<Badge />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Performance Testing', () => {
    it('should handle rapid re-renders efficiently', () => {
      const { rerender } = render(<Badge>Initial Badge</Badge>);

      expect(screen.getByText('Initial Badge')).toBeInTheDocument();

      // Rerender with different content
      rerender(<Badge>Updated Badge</Badge>);

      expect(screen.getByText('Updated Badge')).toBeInTheDocument();
      expect(screen.queryByText('Initial Badge')).not.toBeInTheDocument();
    });

    it('should handle variant changes efficiently', () => {
      const { rerender } = render(<Badge variant="default">Default Variant</Badge>);

      const badge = screen.getByText('Default Variant');
      expect(badge).toHaveClass('border-transparent', 'bg-primary', 'text-primary-foreground');

      // Change to secondary variant
      rerender(<Badge variant="secondary">Secondary Variant</Badge>);

      expect(badge).toHaveClass('border-transparent', 'bg-secondary', 'text-secondary-foreground');
      expect(badge).not.toHaveClass('bg-primary', 'text-primary-foreground');
    });

    it('should handle multiple badge renders efficiently', () => {
      const badges = Array.from({ length: 100 }, (_, i) => `Badge ${i + 1}`);

      const startTime = performance.now();

      render(
        <div>
          {badges.map((text, index) => (
            <Badge key={index}>{text}</Badge>
          ))}
        </div>
      );

      const endTime = performance.now();

      expect(screen.getByText('Badge 1')).toBeInTheDocument();
      expect(screen.getByText('Badge 100')).toBeInTheDocument();
      expect(endTime - startTime).toBeLessThan(1000); // Should render within 1 second
    });
  });
}); 