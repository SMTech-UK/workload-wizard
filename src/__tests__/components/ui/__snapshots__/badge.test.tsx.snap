// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for default badge 1`] = `
<div
  class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-primary text-primary-foreground shadow hover:bg-primary/80"
>
  Default Badge
</div>
`;

exports[`Badge Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for destructive badge 1`] = `
<div
  class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-destructive text-destructive-foreground shadow hover:bg-destructive/80"
>
  Destructive Badge
</div>
`;

exports[`Badge Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for empty badge 1`] = `
<div
  class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-primary text-primary-foreground shadow hover:bg-primary/80"
/>
`;

exports[`Badge Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for outline badge 1`] = `
<div
  class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground"
>
  Outline Badge
</div>
`;

exports[`Badge Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for secondary badge 1`] = `
<div
  class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
>
  Secondary Badge
</div>
`;
